File: trunk/cloudfusion/pyfusebox/virtualfile.py
'''
Created on 24.08.2011

@author: joe
'''
import os
from cloudfusion.pyfusebox.pyfusebox import zstat
import stat
  
class VirtualFile(object):
    INITIAL_TEXT="""
Some virtual Text.
"""
    def __init__(self, path):
        self.path = path
        self.stats = zstat()
        self.text = self.INITIAL_TEXT
        self.stats['st_mode'] = 0777 | stat.S_IFREG
    def getattr(self):
        self.stats['st_size'] = len(self.text)
        self.stats['st_blocks'] = (int) ((self.stats['st_size'] + 4095L) / 4096L);
        return self.stats
    def truncate(self):
        self.text = ''
    def get_text(self):
        return self.text
    def get_path(self):
        return self.path
    def __str__(self):
        return self.path
    def read(self, size, offset):
        return self.text[offset: offset+size]
    def write(self, buf, offset):
        self.text_tmp = self.text[:offset]+buf+self.text[len(buf)+offset:] 
        return len(buf)
    def get_dir(self):
        return os.path.dirname(self.path)
    def get_name(self):
        return os.path.basename(self.path)

File: splitter/src/com/github/joe42/splitter/vtf/VirtualFile.java
package com.github.joe42.splitter.vtf;

import java.io.File;
import java.io.UnsupportedEncodingException;
import java.nio.*;

import com.github.joe42.splitter.util.StringUtil;

import fuse.compat.FuseStat;

public class VirtualFile {
	private String path, text;
	private FileEntry entry;
	public VirtualFile(String path) {
		this.path = path;
		this.text = "";
		entry = new FileEntry();
	}
	protected FileEntry getFileEntry(){
		return entry;
	}

	public FuseStat getAttr(){
		entry.size = getNumOfBytes();
		return entry.getFuseStat();
	}
	private int getNumOfBytes() {
		try {
			return text.getBytes("UTF-8").length;
		} catch (UnsupportedEncodingException e) {
			e.printStackTrace();
		}
		return 0;
	}

	public void truncate(){
		text = "";
	}

	public String getText(){
		return text;
	}

	public String getPath() {
		/**
		 * @return the full path to this virtual file
		 **/
		return path;
	}

	public String toString() {
		return text;
	}

	public void read(ByteBuffer buf, long offset) {
		StringUtil.writeUTF8StringToByteBuffer(text, buf);
	}

	public void write(ByteBuffer buf, long offset) {
		text = StringUtil.getUTF8FromByteBuffer(buf); // FIXME: file cannot be
														// larger than 4096
														// bytes
	}

	public void setText(String text){
		this.text = text;
	}

	public String getDir() {
		String parent = new File(path).getParent();
		if(parent == null){
			parent = path;
		}
		return parent;
	}

	public String getName() {
		return new File(path).getName();
	}
}
